Index: librsvg/rsvg-gobject.c
===================================================================
--- librsvg.orig/rsvg-gobject.c	2012-10-18 17:04:40.000000000 +1100
+++ librsvg/rsvg-gobject.c	2012-10-18 17:05:21.456339611 +1100
@@ -356,3 +356,9 @@
                          "flags", flags,
                          NULL);
 }
+
+RsvgHandle *
+rsvg_handle_new_gz (void)
+{
+    return rsvg_handle_new ();
+}
